src/graphics/webgpu/webgpu-graphics-device.js

```diff
@@ -0,0 +1,320 @@
+import { Debug } from '../../core/debug.js';
+
+import {
+    DEVICETYPE_WEBGPU
+} from '../constants.js';
+
+import { GraphicsDevice } from '../graphics-device.js';
+import { RenderTarget } from '../render-target.js';
+
+import { WebgpuVertexBuffer } from './webgpu-vertex-buffer.js';
+import { WebgpuUniformBuffer } from './webgpu-uniform-buffer.js';
+import { WebgpuIndexBuffer } from './webgpu-index-buffer.js';
+import { WebgpuTexture } from './webgpu-texture.js';
+import { WebgpuShader } from './webgpu-shader.js';
+import { WebgpuRenderPipeline } from './webgpu-render-pipeline.js';
+import { WebgpuRenderTarget } from './webgpu-render-target.js';
+import { WebgpuBindGroupFormat } from './webgpu-bind-group-format.js';
+import { WebgpuBindGroup } from './webgpu-bind-group.js';
+import { WebgpuRenderState } from './webgpu-render-state.js';
+
+/** @typedef {import('../render-pass.js').RenderPass} RenderPass */
+
+class WebgpuGraphicsDevice extends GraphicsDevice {
+    /**
+     * The render target represenging the main framebuffer.
+     *
+     * @type {RenderTarget}
+     */
+    frameBuffer;
+
+    /**
+     * Internal representation of the current render state, as requested by the renderer.
+     * In the future this can be completely replaced by a more optimal solution, where
+     * render states are bundled together (DX11 style) and set using a single call.
+     *
+     * @type {WebgpuRenderState}
+     */
+    renderState = new WebgpuRenderState();
+
+    /**
+     * Object responsible for caching and creation of render pipelines.
+     *
+     * @type {WebgpuRenderPipeline}
+     */
+    renderPipeline = new WebgpuRenderPipeline(this);
+
+    /**
+     * Render pipeline currently set on the device.
+     *
+     * @type {GPURenderPipeline}
+     */
+    pipeline;
+
+    /**
+     * An array of bind group formats, based on currently assigned bind groups
+     *
+     * @type {WebgpuBindGroupFormat[]}
+     */
+    bindGroupFormats = [];
+
+    constructor(canvas, options = {}) {
+        super(canvas);
+
+        this.deviceType = DEVICETYPE_WEBGPU;
+        this.precision = 'hiphp';
+        this.maxSamples = 4;
+        this.supportsUniformBuffers = true;
+    }
+
+    async initWebGpu() {
+
+        if (window.navigator.gpu === null) {
```

Comment:
If it is not supported, it will be `undefined` like anything else that doesn't exist:

```suggestion
        if (!window.navigator.gpu) {
```